src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.cs

```diff
@@ -30,21 +32,60 @@ internal partial class NavigateToItemProvider : INavigateToItemProvider2
             _displayFactory = new NavigateToItemDisplayFactory();
         }
 
-        public ISet<string> KindsProvided { get; } = ImmutableHashSet.Create(
-            NavigateToItemKind.Class,
-            NavigateToItemKind.Constant,
-            NavigateToItemKind.Delegate,
-            NavigateToItemKind.Enum,
-            NavigateToItemKind.EnumItem,
-            NavigateToItemKind.Event,
-            NavigateToItemKind.Field,
-            NavigateToItemKind.Interface,
-            NavigateToItemKind.Method,
-            NavigateToItemKind.Module,
-            NavigateToItemKind.Property,
-            NavigateToItemKind.Structure);
-
-        public bool CanFilter => true;
+        ISet<string> INavigateToItemProvider2.KindsProvided => KindsProvided;
+
+        public ImmutableHashSet<string> KindsProvided
+        {
+            get
+            {
+                var result = ImmutableHashSet.Create<string>(StringComparer.Ordinal);
+                foreach (var project in _workspace.CurrentSolution.Projects)
+                {
+                    var navigateToSearchService = project.LanguageServices.GetService<INavigateToSearchService_RemoveInterfaceAboveAndRenameThisAfterInternalsVisibleToUsersUpdate>();
+                    if (navigateToSearchService != null)
+                    {
+                        result = result.Union(navigateToSearchService.KindsProvided);
+                        continue;
+                    }
+                }
+
+                return result;
+            }
+        }
+
+        public bool CanFilter
+        {
+            get
+            {
+                foreach (var project in _workspace.CurrentSolution.Projects)
+                {
+                    var navigateToSearchService = project.LanguageServices.GetService<INavigateToSearchService_RemoveInterfaceAboveAndRenameThisAfterInternalsVisibleToUsersUpdate>();
+                    if (navigateToSearchService != null)
+                    {
+                        if (!navigateToSearchService.CanFilter)
+                        {
+                            return false;
+                        }
+
+                        continue;
+                    }
+
+#pragma warning disable CS0618 // Type or member is obsolete
+                    var legacyNavigateToSearchService = project.LanguageServices.GetService<INavigateToSearchService>();
```

Comment:
Any reason this method isn't just using your other helper that already does the "return an interface"? I recognize it'll allocate a shim which will then cause this loop to bail, but it's a single allocation per search, which seems pretty cheap compared to the duplicate logic here.

Reply:
(bonus points for a "get all services" helper that just does the loop and is consumed in both properties)